pkg/policy/validate.go

```diff
@@ -316,18 +319,183 @@ func Validate(policy *kyverno.ClusterPolicy, client *dclient.Client, mock bool,
 	}
 
 	if !mock {
-		if err := openAPIController.ValidatePolicyFields(p); err != nil {
+		if err := openAPIController.ValidatePolicyFields(*policy); err != nil {
 			return err
 		}
 	} else {
-		if err := openAPIController.ValidatePolicyMutation(p); err != nil {
+		if err := openAPIController.ValidatePolicyMutation(*policy); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func ValidateVariables(p *kyverno.ClusterPolicy, backgroundMode bool) error {
+	vars := hasVariables(p)
+	if len(vars) == 0 {
+		return nil
+	}
+
+	if err := hasInvalidVariables(p, backgroundMode); err != nil {
+		return fmt.Errorf("policy contains invalid variables: %s", err.Error())
+	}
+
+	if backgroundMode {
+		if err := containsUserVariables(p, vars); err != nil {
+			return fmt.Errorf("only select variables are allowed in background mode. Set spec.background=false to disable background mode for this policy rule: %s ", err)
+		}
+	}
+
+	return nil
+}
+
+// hasInvalidVariables - checks for unexpected variables in the policy
+func hasInvalidVariables(policy *kyverno.ClusterPolicy, background bool) error {
+	for _, r := range policy.Spec.Rules {
+		ruleCopy := r.DeepCopy()
+
+		if err := ruleForbiddenSectionsHaveVariables(ruleCopy); err != nil {
+			return err
+		}
+
+		// skip variable checks on verifyImages.attestations, as variables in attestations are dynamic
+		for _, vi := range ruleCopy.VerifyImages {
+			for _, a := range vi.Attestations {
+				a.Conditions = nil
+			}
+		}
+
+		ctx := buildContext(ruleCopy, background)
+		if _, err := variables.SubstituteAllInRule(log.Log, ctx, *ruleCopy); !checkNotFoundErr(err) {
+			return fmt.Errorf("variable substitution failed for rule %s: %s", ruleCopy.Name, err.Error())
+		}
+	}
+
+	return nil
+}
+
+// for now forbidden sections are match, exclude and
+func ruleForbiddenSectionsHaveVariables(rule *kyverno.Rule) error {
+	var err error
+
+	err = jsonPatchPathHasVariables(rule.Mutation.PatchesJSON6902)
+	if err != nil {
+		return fmt.Errorf("rule \"%s\" should not have variables in patchesJSON6902 path section", rule.Name)
+	}
+
+	err = objectHasVariables(rule.ExcludeResources)
```

Comment:
Need to check any/all?

Reply:
Can you please elaborate? 

Variables are allowed in any/all, and this logic is checking for sections where variables are not allowed.

Reply:
Ignore me. I was referencing `rule.exclude.any/all` which is included in `rule.ExcludeResources`.